### ISSUE TITLE
[bitnami/kafka] kafka-exporter: Error Init Kafka Client: kafka: client has run out of available brokers to talk to

### ISSUE THREAD

**Issue by PDVJAM** [bitnami/charts#18454]
### Name and Version

bitnami/kafka 24.0.10

### What architecture are you using?

amd64

### What steps will reproduce the bug?

Just enable the kafka-exporter:
```yaml
metrics:
  kafka:
    enabled: true
```
### Are you using any custom parameters or values?

No

### What is the expected behavior?

Kafka Exporter to be running.

### What do you see instead?

Been using the latest kafka chart. 
Kafka exported crashes with the following error:
```
I0816 11:13:57.665635       1 kafka_exporter.go:800] Starting kafka_exporter (version=1.7.0, branch=non-git, revision=non-git)
F0816 11:14:01.869155       1 kafka_exporter.go:901] Error Init Kafka Client: kafka: client has run out of available brokers to talk to: 3 errors occurred:
    * EOF
    * EOF
    * EOF
```

**Comment by corico44** [comment]
Hi @PDVJAM, it seems to be related to this bug #18319. We'll keep you updated once the PR is merged to verify that your bug no longer appears.

**Comment by PDVJAM** [comment]
@corico44 I have made those changes to templates manually, but it didn't help.

**Comment by navi86** [comment]
@corico44 it doesn't look like this pr https://github.com/bitnami/charts/pull/18319.
will help because I also made these changes manually and I am still getting the same error.

**Comment by PDVJAM** [comment]
Any updates? @corico44

**Comment by migruiz4** [comment]
Could you please provide more details about your issue @PDVJAM @navi86? 

I tried to reproduce you issue using the values.yaml in the issue description:
```
metrics:
  kafka:
    enabled: true
```

While using the branch from https://github.com/bitnami/charts/pull/18319, and the exporter successfully connects to Kafka:
```
$ helm install test bitnami/kafka -f values.yaml
...
$ kubectl get pods
NAME                                   READY   STATUS    RESTARTS   AGE
test-kafka-controller-0                1/1     Running   0          9m40s
test-kafka-controller-1                1/1     Running   0          9m40s
test-kafka-controller-2                1/1     Running   0          9m40s
test-kafka-exporter-85dd48f884-7cg96   1/1     Running   0          9m40s
$ kubectl logs test-kafka-exporter-85dd48f884-7cg96
I0825 13:11:03.447521       1 kafka_exporter.go:800] Starting kafka_exporter (version=1.7.0, branch=non-git, revision=non-git)
I0825 13:11:03.486450       1 kafka_exporter.go:971] Listening on HTTP :9308
$ kubectl port-forward svc/test-kafka-metrics 9308:9308
Forwarding from 127.0.0.1:9308 -> 9308
Forwarding from [::1]:9308 -> 9308
$ curl localhost:9308/metrics
...
# HELP kafka_broker_info Information about the Kafka Broker.
# TYPE kafka_broker_info gauge
kafka_broker_info{address="test-kafka-controller-0.test-kafka-controller-headless.default.svc.cluster.local:9092",id="0"} 1
kafka_broker_info{address="test-kafka-controller-1.test-kafka-controller-headless.default.svc.cluster.local:9092",id="1"} 1
kafka_broker_info{address="test-kafka-controller-2.test-kafka-controller-headless.default.svc.cluster.local:9092",id="2"} 1
...

```

**Comment by PDVJAM** [comment]
I literally copied the updated deployment.yaml from that branch and it didn't work for me:
```
I0825 14:13:23.989790       1 kafka_exporter.go:800] Starting kafka_exporter (version=1.7.0, branch=non-git, revision=non-git)
F0825 14:13:28.411938       1 kafka_exporter.go:901] Error Init Kafka Client: kafka: client has run out of available brokers to talk to: 3 errors occurred:
	* EOF
	* EOF
	* EOF
```

deployment.yaml:
```
{{- /*
Copyright VMware, Inc.
SPDX-License-Identifier: APACHE-2.0
*/}}

{{- if .Values.metrics.kafka.enabled }}
{{- $releaseNamespace := include "common.names.namespace" . -}}
{{- $clusterDomain := .Values.clusterDomain -}}
{{- $fullname := include "common.names.fullname" . -}}
{{- $containerPort := int .Values.listeners.client.containerPort -}}
apiVersion: {{ include "common.capabilities.deployment.apiVersion" . }}
kind: Deployment
metadata:
  name: {{ include "kafka.metrics.kafka.fullname" . }}
  namespace: {{ include "common.names.namespace" . | quote }}
  labels: {{- include "common.labels.standard" ( dict "customLabels" .Values.commonLabels "context" $ ) | nindent 4 }}
    app.kubernetes.io/component: cluster-metrics
  {{- if .Values.commonAnnotations }}
  annotations: {{- include "common.tplvalues.render" ( dict "value" .Values.commonAnnotations "context" $ ) | nindent 4 }}
  {{- end }}
spec:
  replicas: 1
  {{- $podLabels := merge .Values.metrics.kafka.podLabels .Values.commonLabels }}
  selector:
    matchLabels: {{- include "common.labels.matchLabels" ( dict "customLabels" $podLabels "context" $ ) | nindent 6 }}
      app.kubernetes.io/component: cluster-metrics
  template:
    metadata:
      labels: {{- include "common.labels.standard" ( dict "customLabels" $podLabels "context" $ ) | nindent 8 }}
        app.kubernetes.io/component: cluster-metrics
      {{- if .Values.metrics.kafka.podAnnotations }}
      annotations: {{- include "common.tplvalues.render" (dict "value" .Values.metrics.kafka.podAnnotations "context" $) | nindent 8 }}
      {{- end }}
    spec:
      {{- include "kafka.imagePullSecrets" . | nindent 6 }}
      {{- if .Values.metrics.kafka.hostAliases }}
      hostAliases: {{- include "common.tplvalues.render" (dict "value" .Values.metrics.kafka.hostAliases "context" $) | nindent 8 }}
      {{- end }}
      {{- if .Values.metrics.kafka.affinity }}
      affinity: {{- include "common.tplvalues.render" ( dict "value" .Values.metrics.kafka.affinity "context" $) | nindent 8 }}
      {{- else }}
      affinity:
        podAffinity: {{- include "common.affinities.pods" (dict "type" .Values.metrics.kafka.podAffinityPreset "component" "metrics"  "customLabels" $podLabels "context" $) | nindent 10 }}
        podAntiAffinity: {{- include "common.affinities.pods" (dict "type" .Values.metrics.kafka.podAntiAffinityPreset "component" "metrics"  "customLabels" $podLabels "context" $) | nindent 10 }}
        nodeAffinity: {{- include "common.affinities.nodes" (dict "type" .Values.metrics.kafka.nodeAffinityPreset.type "key" .Values.metrics.kafka.nodeAffinityPreset.key "values" .Values.metrics.kafka.nodeAffinityPreset.values) | nindent 10 }}
      {{- end }}
      {{- if .Values.metrics.kafka.nodeSelector }}
      nodeSelector: {{- include "common.tplvalues.render" ( dict "value" .Values.metrics.kafka.nodeSelector "context" $) | nindent 8 }}
      {{- end }}
      {{- if .Values.metrics.kafka.tolerations }}
      tolerations: {{- include "common.tplvalues.render" (dict "value" .Values.metrics.kafka.tolerations "context" .) | nindent 8 }}
      {{- end }}
      {{- if .Values.metrics.kafka.topologySpreadConstraints }}
      topologySpreadConstraints: {{- include "common.tplvalues.render" (dict "value" .Values.metrics.kafka.topologySpreadConstraints "context" .) | nindent 8 }}
      {{- end }}
      {{- if .Values.metrics.kafka.priorityClassName }}
      priorityClassName: {{ .Values.metrics.kafka.priorityClassName }}
      {{- end }}
      {{- if .Values.metrics.kafka.schedulerName }}
      schedulerName: {{ .Values.metrics.kafka.schedulerName }}
      {{- end }}
      {{- if .Values.metrics.kafka.podSecurityContext.enabled }}
      securityContext: {{- omit .Values.metrics.kafka.podSecurityContext "enabled" | toYaml | nindent 8 }}
      {{- end }}
      serviceAccountName: {{ template "kafka.metrics.kafka.serviceAccountName" . }}
      enableServiceLinks: {{ .Values.metrics.kafka.enableServiceLinks }}
      {{- if .Values.metrics.kafka.initContainers }}
      initContainers: {{- include "common.tplvalues.render" (dict "value" .Values.metrics.kafka.initContainers "context" $) | nindent 8 }}
      {{- end }}
      containers:
        - name: kafka-exporter
          image: {{ include "kafka.metrics.kafka.image" . }}
          imagePullPolicy: {{ .Values.metrics.kafka.image.pullPolicy | quote }}
          {{- if .Values.metrics.kafka.containerSecurityContext.enabled }}
          securityContext: {{- omit .Values.metrics.kafka.containerSecurityContext "enabled" | toYaml | nindent 12 }}
          {{- end }}
          {{- if .Values.diagnosticMode.enabled }}
          command: {{- include "common.tplvalues.render" (dict "value" .Values.diagnosticMode.command "context" $) | nindent 12 }}
          {{- else if .Values.metrics.kafka.command }}
          command: {{- include "common.tplvalues.render" (dict "value" .Values.metrics.kafka.command "context" $) | nindent 12 }}
          {{- else }}
          command:
            - bash
          {{- end }}
          {{- if .Values.diagnosticMode.enabled }}
          args: {{- include "common.tplvalues.render" (dict "value" .Values.diagnosticMode.args "context" $) | nindent 12 }}
          {{- else if .Values.metrics.kafka.args }}
          args: {{- include "common.tplvalues.render" (dict "value" .Values.metrics.kafka.args "context" $) | nindent 12 }}
          {{- else }}
          args:
            - -ce
            - |
              kafka_exporter \
              {{- range $i := until (int .Values.controller.replicaCount) }}
              --kafka.server={{ $fullname }}-controller-{{ $i }}.{{ $fullname }}-controller-headless.{{ $releaseNamespace }}.svc.{{ $clusterDomain }}:{{ $containerPort }} \
              {{- end }}
              {{- range $i := until (int .Values.broker.replicaCount) }}
              --kafka.server={{ $fullname }}-broker-{{ $i }}.{{ $fullname }}-broker-headless.{{ $releaseNamespace }}.svc.{{ $clusterDomain }}:{{ $containerPort }} \
              {{- end }}
              {{- if regexFind "SASL" (upper .Values.listeners.client.protocol) }}
              --sasl.enabled \
              --sasl.username=$SASL_USERNAME \
              --sasl.password=$SASL_USER_PASSWORD \
              --sasl.mechanism={{ include "kafka.metrics.kafka.saslMechanism" . }} \
              {{- end }}
              {{- if regexFind "SSL" (upper .Values.listeners.client.protocol) }}
              --tls.enabled \
              {{- if .Values.metrics.kafka.certificatesSecret }}
              --tls.key-file=/opt/bitnami/kafka-exporter/certs/{{ .Values.metrics.kafka.tlsKey }} \
              --tls.cert-file=/opt/bitnami/kafka-exporter/certs/{{ .Values.metrics.kafka.tlsCert }} \
              {{- if .Values.metrics.kafka.tlsCaSecret }}
              --tls.ca-file=/opt/bitnami/kafka-exporter/cacert/{{ .Values.metrics.kafka.tlsCaCert }} \
              {{- else }}
              --tls.ca-file=/opt/bitnami/kafka-exporter/certs/{{ .Values.metrics.kafka.tlsCaCert }} \
              {{- end }}
              {{- end }}
              {{- end }}
              {{- range $key, $value := .Values.metrics.kafka.extraFlags }}
              --{{ $key }}{{ if $value }}={{ $value }}{{ end }} \
              {{- end }}
              --web.listen-address=:{{ .Values.metrics.kafka.containerPorts.metrics }}
          {{- end }}
          {{- if regexFind "SASL" (upper .Values.listeners.client.protocol) }}
          env:
            - name: SASL_USERNAME
              value: {{ index .Values.sasl.client.users 0 | quote }}
            - name: SASL_USER_PASSWORD
              valueFrom:
                secretKeyRef:
                  name: {{ include "kafka.saslSecretName" . }}
                  key: system-user-password
            {{- end }}
          ports:
            - name: metrics
              containerPort: {{ .Values.metrics.kafka.containerPorts.metrics }}
          {{- if .Values.metrics.kafka.resources }}
          resources: {{ toYaml .Values.metrics.kafka.resources | nindent 12 }}
          {{- end }}
          volumeMounts:
            {{- if .Values.metrics.kafka.extraVolumeMounts }}
            {{- include "common.tplvalues.render" (dict "value" .Values.metrics.kafka.extraVolumeMounts "context" $) | nindent 12 }}
            {{- end }}
            {{- if and (regexFind "SSL" (upper .Values.listeners.client.protocol)) .Values.metrics.kafka.certificatesSecret }}
            - name: kafka-exporter-certificates
              mountPath: /opt/bitnami/kafka-exporter/certs/
              readOnly: true
            {{- if .Values.metrics.kafka.tlsCaSecret }}
            - name: kafka-exporter-ca-certificate
              mountPath: /opt/bitnami/kafka-exporter/cacert/
              readOnly: true
            {{- end }}
            {{- end }}
        {{- if .Values.metrics.kafka.sidecars }}
        {{- include "common.tplvalues.render" (dict "value" .Values.metrics.kafka.sidecars "context" $) | nindent 8 }}
        {{- end }}
      volumes:
        {{- if .Values.metrics.kafka.extraVolumes }}
        {{- include "common.tplvalues.render" (dict "value" .Values.metrics.kafka.extraVolumes "context" $) | nindent 8 }}
        {{- end }}
        {{- if and (regexFind "SSL" (upper .Values.listeners.client.protocol)) .Values.metrics.kafka.certificatesSecret }}
        - name: kafka-exporter-certificates
          secret:
            secretName: {{ .Values.metrics.kafka.certificatesSecret }}
            defaultMode: 0440
        {{- if .Values.metrics.kafka.tlsCaSecret }}
        - name: kafka-exporter-ca-certificate
          secret:
            secretName: {{ .Values.metrics.kafka.tlsCaSecret }}
            defaultMode: 0440
        {{- end }}
        {{- end }}
{{- end }}
```

**Comment by PDVJAM** [comment]
I have deployed latest chart: 25.1.2.
It is all the same:
```
k -n dev logs kafka-exporter-77f75855d-2kmzl -f
I0829 09:21:17.221513       1 kafka_exporter.go:800] Starting kafka_exporter (version=1.7.0, branch=non-git, revision=non-git)
F0829 09:21:21.660143       1 kafka_exporter.go:901] Error Init Kafka Client: kafka: client has run out of available brokers to talk to: 3 errors occurred:
	* EOF
	* EOF
	* EOF
```

**Comment by PDVJAM** [comment]
Okay, I just checked what @migruiz4 did. It works with a clean install.
```
helm install test bitnami/kafka -f values.yaml --namespace=default                                                                       15s ○ do-sfo2-nft-k8s-cluster
NAME: test
LAST DEPLOYED: Wed Aug 30 18:11:11 2023
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
CHART NAME: kafka
CHART VERSION: 25.1.4
APP VERSION: 3.5.1
...
k logs test-kafka-exporter-5d5d66c7bd-vdpjn -f
I0830 16:12:28.485519       1 kafka_exporter.go:800] Starting kafka_exporter (version=1.7.0, branch=non-git, revision=non-git)
I0830 16:12:28.544466       1 kafka_exporter.go:971] Listening on HTTP :9308
```

But if I try to enable it with the already running cluster by upgrading it, it does not. Maybe there is something on persistent volumes, some kind of old, broken configuration?
```
helm upgrade --install --namespace=dev kafka bitnami/kafka -f ./kafka/dev.yaml
...
k -n dev logs kafka-exporter-5c4b6d69b4-6f6b6 -f
I0830 16:50:22.681457       1 kafka_exporter.go:800] Starting kafka_exporter (version=1.7.0, branch=non-git, revision=non-git)
F0830 16:50:27.098816       1 kafka_exporter.go:901] Error Init Kafka Client: kafka: client has run out of available brokers to talk to: 3 errors occurred:
	* EOF
	* EOF
	* EOF
```

**Comment by PDVJAM** [comment]
Okay, it seems that the problem is with auth. 
```
    Args:
      -ce
      kafka_exporter \
      --kafka.server=kafka-controller-0.kafka-controller-headless.dev.svc.cluster.local:9092 \
      --kafka.server=kafka-controller-1.kafka-controller-headless.dev.svc.cluster.local:9092 \
      --kafka.server=kafka-controller-2.kafka-controller-headless.dev.svc.cluster.local:9092 \
      --sasl.enabled \
      --sasl.username=$SASL_USERNAME \
      --sasl.password=$SASL_USER_PASSWORD \
      --sasl.mechanism=plain \
      --web.listen-address=:9308

...
    Environment:
      SASL_USERNAME:       infra
      SASL_USER_PASSWORD:  <set to the key 'system-user-password' in secret 'kafka-auth'>  Optional: false
```

infra is the first user from the client users:
```
sasl:
  enabledMechanisms: PLAIN
  client:
    users:
      - infra
      - worker
      - api
      - producer
  existingSecret: "kafka-auth"
```
Its password is the first one in the secret kafka-auth, key client-passwords. But the exporter wants to use system-user-password, which is actually not described anywhere at all.

@corico44 @migruiz4

**Comment by PDVJAM** [comment]
@corico44 @migruiz4 Any updates?

**Comment by eli-gc** [comment]
I am experiencing the same issue for chart version: 25.1.10

**Comment by eli-gc** [comment]
The cause of my error was actually different. I made an separate issue #19457 for it.
@PDVJAM You should try with the default user: "user1". That user worked with the default system-user-password for me. Might help you narrow down your issue.

**Comment by PDVJAM** [comment]
Okay, the issue is that system-user-password must be the 0 client user password. Probably would be good to mention that somewhere in the documentation?..

**Comment by PDVJAM** [comment]
> so what should be done @PDVJAM ? I just got the same issue
> 
The first password in the client-passwords is the system user password being used for the exported needs.

**Comment by dmitry-mightydevops** [comment]
@PDVJAM 
can you elaborate? I don't quite understand.

**Comment by PDVJAM** [comment]
> @PDVJAM can you elaborate? I don't quite understand.
> 
> ```shell
> ➜ k ice env kafka-exporter-7fb5c5dcdb-r5vrj                                      
> CONTAINER       NAME                VALUE
> kafka-exporter  SASL_USERNAME       user1
> kafka-exporter  SASL_USER_PASSWORD  SECRETMAP:kafka-user-passwords KEY:system-user-password
> ```
> 
> so this feels like a correct setup, as generated `kafka-user-passwords` has same value for the following keys of the secret:
> 
>     * client-passwords
> 
>     * system-user-password
> 
> 
> ```shell
> ➜ kgsec kafka-user-passwords -o go-template='{{range $k,$v := .data}}{{printf "%s: " $k}}{{if not $v}}{{$v}}{{else}}{{$v | base64decode}}{{end}}{{"\n"}}{{end}}'
> client-passwords: 05dRAwhtKX
> controller-password: 3VT0vYYV5l
> inter-broker-password: 7iiXMZzELS
> system-user-password: 05dRAwhtKX
> ```

For me it started to work when I did this:
```
sasl:
  enabledMechanisms: PLAIN
  client:
    users:
      - user0
      - user1
      - user2
      - user3
      - user4
  existingSecret: "kafka-auth"
```

```
kubedecode kafka-auth dev
client-passwords: user0pw,user1pw,user2pw,user3pw,user4pw
```

```
k -n dev exec -it kafka-exporter-6b4475ffdc-64jsg -- env | grep SASL
SASL_USER_PASSWORD=user0pw
SASL_USERNAME=user0
```

It doesn't use system-user-password AFAIR.

**Comment by PDVJAM** [comment]
> we tried setting system-user-password and the client-passwords as the same string, and still get the same issue.

It doesn't use system-user-password, it uses client-password.0 for that.
